GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m

```diff
@@ -223,21 +223,27 @@ - (GDTCORBackgroundIdentifier)beginBackgroundTaskWithName:(NSString *)name
   }
 #endif  // !NDEBUG
   return bgID;
-#endif  // !TARGET_OS_WATCH
-  // TODO: WKExtension background tasks handling.
-  return GDTCORBackgroundIdentifierInvalid;
 }
 
 - (void)endBackgroundTask:(GDTCORBackgroundIdentifier)bgID {
-#if !TARGET_OS_WATCH
   if (bgID != GDTCORBackgroundIdentifierInvalid) {
     GDTCORLogDebug("Ending background task with ID:%ld was successful", (long)bgID);
     [[self sharedApplicationForBackgroundTask] endBackgroundTask:bgID];
     return;
   }
-#endif  // !TARGET_OS_WATCH
 }
 
+#elif TARGET_OS_WATCH
+- (id<NSObject>)beginActivityWithOptions:(NSActivityOptions)options reason:(NSString *)reason {
+  return [[self sharedNSProcessInfoForBackgroundTask] beginActivityWithOptions:options
```

Comment:
From the [watchOS lifecycle docs](https://developer.apple.com/documentation/watchkit/working_with_the_watchos_app_life_cycle?language=objc)
> Because the system can purge suspended apps without warning, use the extension delegate’s applicationDidEnterBackground method to save any data you need to recreate your app’s current state. If needed, you can request additional background execution time by calling the NSProcessInfo class’s performExpiringActivityWithReason:usingBlock: method.

That might mean that one possible solution is to just call the appWillTerminate callbacks when receiving a backgrounding notice on watchOS. Another might be implementing the handleBackgroundTasks method.

It also means that this might not be the right API to call--you might need to call `performExpiringActivityWithReason:usingBlock:` instead. If that's the case, it's not really possible to know how long we'll need to execute. It might be sufficient to just write a line in there that spins the current thread loop for like 10 seconds or something.

My hunch is that we'll want to call `performExpiringActivityWithReason:usingBlock:` and simultaneously call the `appWillTerminate` callbacks to try and force each queue to finish its work. @chliangGoogle what do you think?